File: M3 Intermediate DSA/src/L5_Arrays_Prefix_Sum/Assignment5.java
package L5_Arrays_Prefix_Sum;

public class Assignment5 {

  public int solve(int[] A) {
    int N = A.length;
    int[] prefixSum = new int[N];

    prefixSum[0] = A[0];

    for (int index = 1; index < N; index++) {
      prefixSum[index] = prefixSum[index - 1] + A[index];
    }

    if (prefixSum[N - 1] - prefixSum[0] == 0) {
      return 0;
    }

    for (int index = 1; index < N ;index++) {
      int leftSum = prefixSum[index -1];
      int rightSum = prefixSum[N-1] - prefixSum[index];

      if (leftSum == rightSum) {
        return index;
      }
    }

    return -1;
  }

}

File: M3_Intermediate_DSA/L3_Prefix_Sum/assignment5.go
package main

func calPrefixSum(A []int) []int {
	N := len(A)

	prefixSum := make([]int, N)
	prefixSum[0] = A[0]

	for index := 1; index < N; index++ {
		prefixSum[index] = prefixSum[index-1] + A[index]
	}

	return prefixSum
}

func SolveA5(A []int) int {
	N := len(A)
	prefixSum := calPrefixSum(A)

	if prefixSum[N-1]-prefixSum[0] == 0 {
		return 0
	}

	for index := 1; index < N; index++ {
		leftSum := prefixSum[index-1]
		rightSum := prefixSum[N-1] - prefixSum[index]

		if leftSum == rightSum {
			return index
		}
	}

	return -1
}
